### PR TITLE
Feat/63 registrere enkeltdelegering av ressurs

### PR DESCRIPTION
**Pull request by Sondre-B**
Åpnet opp for å kunne registrere enkeltdelegering av en ressurs fra ResourceRegistry.

## Description
Lagt til operasjonen AddResourceDelegation for å kunne registrere enkeltdelegering for en ressurs.

## Related Issue(s)
- #63

### CODE REVIEW

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/DelegationHelper.cs [hunk 1]
```diff
@@ -20,12 +17,12 @@ public static class DelegationHelper
         /// <param name="rules">The list of rules to be sorted</param>
         /// <param name="unsortableRules">The list of rules not able to sort by org/app/offeredBy/CoveredBy</param>
         /// <returns>A dictionary with key being the filepath for the delegation policy file, and value being the list of rules to be written to the delegation policy</returns>
-        public static Dictionary<string, List<PolicyRule>> SortRulesByDelegationPolicyPath(List<PolicyRule> rules, out List<PolicyRule> unsortableRules)
+        public static Dictionary<string, List<Models.Rule>> SortRulesByDelegationPolicyPath(List<Models.Rule> rules, out List<Models.Rule> unsortableRules)
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/DelegationHelper.cs:20]: Models. in object type an be removed by adding using statement

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql [hunk 2]
```diff
@@ -1,22 +1,22 @@
...
+	_delegationchangetype delegation.delegationchangetype,
+	_altinnappid character varying,
+	_offeredbypartyid integer,
+	_coveredbyuserid integer,
+	_coveredbypartyid integer,
+	_performedbyuserid integer,
+	_blobstoragepolicypath character varying,
+	_blobstorageversionid character varying,
+	_resourceid character varying,
+	_resourcetype character varying)
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql:12]: _resourcetype is defined as a Type in the resourceregistry schema. We may not be able to reference it directly but should in stead make our own Type in the delegation schema?

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql [hunk 3]
```diff
@@ -1,22 +1,22 @@
...
+    RETURNS SETOF delegation.delegationchanges 
+    LANGUAGE 'sql'
+    VOLATILE
+    ROWS 1
+
 AS $BODY$
-  INSERT INTO delegation.delegationChanges(
+INSERT INTO delegation.delegationChanges(
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql:19]: Missing 01-setup-tables.sql script for actually adding the columns to the delegationChanges table?

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs [hunk 4]
```diff
@@ -202,7 +202,7 @@ public static string GetAltinnAppDelegationPolicyPath(string org, string app, st
             }
             else
             {
-                return $"resourceregistry /{resourceRegistryId}/{offeredBy}/{coveredBy}/ delegationpolicy.xacml";
+                return $"resourceregistry/{resourceRegistryId}/{offeredBy}/{coveredBy}/resourcepolicy.xml";
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:205]: The delegation policy should still be called delegationpolicy.xml even if it's a delegation of a resource from the registry.
Otherwise the change in PolicyRepository.CreateBlobClient which use the policy name to determine which storage account to use will fail.

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Persistance/DelegationMetadataRepository.cs [hunk 5]
```diff
@@ -76,15 +77,24 @@ public async Task<DelegationChange> InsertDelegation(DelegationChange delegation
...
-
-                NpgsqlCommand pgcom = new NpgsqlCommand(getCurrentDelegationChangeSql, conn);
-                pgcom.Parameters.AddWithValue("_altinnAppId", altinnAppId);
+                NpgsqlCommand pgcom = new NpgsqlCommand();
+                if (!string.IsNullOrWhiteSpace(resourceRegistryId))
+                {
+                    pgcom = new NpgsqlCommand(getCurrentDelegationChangeBasedOnResourceRegistryIdSql, conn);
+                    pgcom.Parameters.AddWithValue("_resourceregistryid", resourceRegistryId);
+                }
+                else if (!string.IsNullOrWhiteSpace(altinnAppId) || altinnAppId != "/")
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Persistance/DelegationMetadataRepository.cs:92]: if the altinnAppId is ever "/" then there is an earlier bug in the code which still tries to build an appId without Org and App being the resource identifiers. So why is this needed?

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/appsettings.json [hunk 6]
```diff
@@ -23,17 +23,31 @@
...
-    "MetadataBlobEndpoint": "http://127.0.0.1:10000/devstoreaccount1",
+    "MetadataBlobEndpoint": "https://pocaccessmanagement.blob.core.windows.net/",
     "DelegationsAccountName": "devstoreaccount1",
     "DelegationsAccountKey": "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==",
     "DelegationsContainer": "delegationpolicies",
-    "DelegationsBlobEndpoint": "http://127.0.0.1:10000/devstoreaccount1",
-    "BlobLeaseTimeout": 15,
+    "DelegationsBlobEndpoint": "https://pocaccessmanagement.blob.core.windows.net/",
+    "ResourceRegistryAccountName": "pocaccessmanagement",
+    "ResourceRegistryAccountKey": "wiA0t9w4jYCn56f4MIfKLRf3Nm+Srg1d8W+9vNJhRPtJUVo+3Q+zC+uWCg3StPHb9QhIJoiu7qWU+ASt2gHQ0Q==",
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/appsettings.json:32]: You should change your access key now :P
and reset the appsettings.json you commit to the default devstoreaccount1 as the others

Comment (jonkjetiloye) [reply backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/appsettings.json:32]: also for MetadataBlobEndpoint

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql [hunk 7]
```diff
@@ -37,7 +37,48 @@ AS $BODY$
...
     _blobStorageVersionId,
-    _resourceid,
-    _resourcetype
+	_resourceid,
+	_resourcetype
   ) RETURNING *;
 $BODY$;
+
+ALTER FUNCTION delegation.insert_delegationchange(delegation.delegationchangetype, character varying, integer, integer, integer, integer, character varying, character varying, character varying, character varying)
+    OWNER TO platform_authorization_admin;
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql:46]: Redundant? platform_authorization_admin is the user running these script and creates the database artifacts. 
If, anything you might need to run grant script for new functions to the runtime platform_authorization user

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 8]
```diff
@@ -151,12 +151,7 @@ private async Task<bool> WriteDelegationPolicyInternal(string policyPath, List<P
                     return false;
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:151]: Logger warning must be updated and appPolicy variable should be renamed resourcePolicy

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Controllers/DelegationsController.cs [hunk 9]
```diff
@@ -82,7 +81,7 @@ public async Task<ActionResult> Post([FromBody] List<PolicyRule> rules)
         /// <response code="500">Internal Server Error</response>
         [HttpPost]
         [Route("authorization/api/v1/[controller]/AddResourceDelegation")]
-        public async Task<ActionResult> AddResourceDelegation([FromBody] List<PolicyRule> rules)
+        public async Task<ActionResult> AddResourceDelegation([FromBody] List<Rule> rules)
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Controllers/DelegationsController.cs:84]: AddResourceDelegation is now just AddRules without Authorization.
We will need a endpoint with user authentication and authorization at some point before frontend can connect.

If we are not also adding the new authentication/authorization for end users together with this change I suggest we remove the endpoint, and just test through AddRules

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql (hunk 3 above)
Comment (Sondre-B) [reply backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Migration/v0.07-authorization-admin/02-setup-procedures.sql:19]: No changes in 01-setup-tables.sql in this PR but the script exists in repo.

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Altinn.AuthorizationAdmin.Core.csproj [hunk 10]
```diff
@@ -10,6 +10,7 @@
     <PackageReference Include="Azure.Storage.Blobs" Version="12.13.1" />
     <PackageReference Include="Azure.Storage.Queues" Version="12.11.1" />
     <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="6.0.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Altinn.AuthorizationAdmin.Core.csproj:13]: There's been a decision to move away from dependencies to Newtonsoft.Json. Use System.Text.Json instead

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs [hunk 11]
```diff
@@ -108,7 +108,7 @@ public static string GetAltinnAppsPolicyPath(string resourceRegistryId)
                 throw new ArgumentException("ResourceRegistryId was not defined");
             }
 
-            return $"{resourceRegistryId.AsFileName()}/policy.xml";
+            return $"{resourceRegistryId.AsFileName()}/resourcepolicy.xml";
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:111]: The function name should be renamed from GetAltinnAppsPolicyPath to GetResourceRegistryPolicyPath.
Documentation of resourceRegistryId property also needs update

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs [hunk 12]
```diff
@@ -141,6 +141,7 @@ public static Rule CreateRuleFromPolicyAndRuleMatch(RequestToDelete search, Xacm
         /// <param name="offeredBy">The party id of the entity offering the delegated the policy</param>
         /// <param name="coveredByUserId">The user id of the entity having received the delegated policy or null if party id</param>
         /// <param name="coveredByPartyId">The party id of the entity having received the delegated policy or null if user id</param>
+        /// <param name="resourceRegistryId">The id of the resource in the resourceregistry.</param>
         /// <returns>policypath matching input data</returns>
         public static string GetAltinnAppDelegationPolicyPath(string org, string app, string offeredBy, int? coveredByUserId, int? coveredByPartyId, string resourceRegistryId)
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:146]: Method should be renamed from GetAltinnAppDelegationPolicyPath to GetDelegationPolicyPath as it is no longer altinn app specific

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Models/ResourceRegistry/ServiceResource.cs [hunk 13]
```diff
@@ -1,4 +1,8 @@
-﻿using System.Text.Json.Serialization;
+﻿using System.Runtime.Serialization;
+using System.Text.Json.Serialization;
+using Newtonsoft.Json;
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Models/ResourceRegistry/ServiceResource.cs:3]: Same here regarding Newtonsoft.Json; dependency

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Persistance/DelegationMetadataRepository.cs [hunk 14]
```diff
@@ -57,6 +58,8 @@ public async Task<DelegationChange> InsertDelegation(DelegationChange delegation
                 pgcom.Parameters.AddWithValue("_performedByUserId", delegationChange.PerformedByUserId);
                 pgcom.Parameters.AddWithValue("_blobStoragePolicyPath", delegationChange.BlobStoragePolicyPath);
                 pgcom.Parameters.AddWithValue("_blobStorageVersionId", delegationChange.BlobStorageVersionId);
+                pgcom.Parameters.AddWithValue("_resourceid", delegationChange.ResourceId);
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Persistance/DelegationMetadataRepository.cs:61]: ResourceId and ResourceType columns are now not created as nullable in 01-setup-tables.sql script.

Won't this cause the previous version of the delegation.insert_delegationchange function (which does not have these properties) to fail?

Also this function now called for an AltinnApp delegation will cause default ResourceType and an empty ResourceId to be inserted?

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/Altinn.AuthorizationAdmin.Tests.csproj [hunk 15]
```diff
@@ -51,6 +51,12 @@
     </Content>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/Altinn.AuthorizationAdmin.Tests.csproj:54]: Is this on purpose?

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Models/ResourceRegistry/ServiceResource.cs [hunk 16]
```diff
@@ -10,86 +14,103 @@ public class ServiceResource
...
         /// ThematicArea
         /// </summary>
+        [JsonProperty]
         public string? ThematicArea { get; set; }
 
         /// <summary>
         /// ResourceReference
         /// </summary>
+        [JsonProperty]
         public ResourceReference? ResourceReference { get; set;  }
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Models/ResourceRegistry/ServiceResource.cs:84]: ResourceReference property is being changed to a List of references in: 
https://github.com/Altinn/altinn-resource-registry/pull/29

Comment (jonkjetiloye) [reply backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Models/ResourceRegistry/ServiceResource.cs:84]: Also updates ReferenceType with two new values which must be added.

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs [hunk 17]
```diff
@@ -202,7 +203,7 @@ public static string GetAltinnAppDelegationPolicyPath(string org, string app, st
             }
             else
             {
-                return $"resourceregistry /{resourceRegistryId}/{offeredBy}/{coveredBy}/ delegationpolicy.xacml";
+                return $"resourceregistry/{resourceRegistryId}/{offeredBy}/{coveredByPrefix}/delegationpolicy.xml";
```
[the remark this answers is not in the document]
Comment (Sondre-B) [reply backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs]: Added check for ResourceRegistryId and added .AsFileName() string extension in registryId scenario.

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs [hunk 18]
```diff
@@ -157,6 +158,11 @@ public static string GetAltinnAppDelegationPolicyPath(string org, string app, st
                 {
                     throw new ArgumentException("App was not defined");
                 }
+
+                if (string.IsNullOrWhiteSpace(org) && string.IsNullOrWhiteSpace(app) && string.IsNullOrWhiteSpace(resourceRegistryId))
+                {
+                    throw new ArgumentException("ResourceRegistryId was not defined");
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:164]: Because of the order of these checks this will never happen as it would already have thrown exception for missing org

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs [hunk 19]
```diff
@@ -202,7 +208,7 @@ public static string GetAltinnAppDelegationPolicyPath(string org, string app, st
             }
             else
             {
-                return $"resourceregistry /{resourceRegistryId}/{offeredBy}/{coveredBy}/ delegationpolicy.xacml";
+                return $"resourceregistry/{resourceRegistryId.AsFileName()}/{offeredBy.AsFileName()}/{coveredByPrefix.AsFileName()}/delegationpolicy.xml";
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:211]: missing the actual coveredBy in path, now only sets the prefix. I'm surprised this doesn't cause everything to fail 🤔

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 20]
```diff
@@ -143,25 +149,31 @@ private async Task<bool> WriteDelegationPolicyInternal(string policyPath, List<R
...
                 {
-                    if (!DelegationHelper.PolicyContainsMatchingRule(appPolicy, rule))
+                    if (!DelegationHelper.PolicyContainsMatchingRule(resourcePolicy, rule))
                     {
-                        _logger.LogWarning("Matching rule not found in app policy. Action might not exist for Resource, or Resource itself might not exist. Delegation policy path: {policyPath}. Rule: {rule}", policyPath, rule);
+                        _logger.LogWarning("Matching rule not found in resource policy. Action might not exist for Resource, or Resource itself might not exist. Delegation policy path: {policyPath}. Rule: {rule}", policyPath, rule);
                         return false;
                     }
+
+                    ServiceResource resource = new ServiceResource();
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:167]: All rules is already sorted into sets for delegations of the same resource (the resourceRegistryId variable).
So shouldnt have to iterate over all rules internally here to check each against _resourceRegistryClient.GetResource.

Instead this can be performed once along side verification that the resourcePolicy exist

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs (hunk 18 above)
Comment (Sondre-B) [reply backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:164]: This is necessary for the code to throw error of missing resourceRegistryId only when resourceRegistryId is empty. Without this check it threw the error also when org and app was set.

Comment (jonkjetiloye) [reply backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Core/Helpers/PolicyHelper.cs:164]: Yes, but now this error will never be thrown because it requires org to be IsNullOrWhiteSpace to be true.
And if so, it will already have thrown the org missing argument.

I'm wondering whether the code should be structured differently for better separation of the org/app vs resourceId flows

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/appsettings.json [hunk 21]
```diff
@@ -23,11 +23,15 @@
     "MetadataAccountName": "devstoreaccount1",
     "MetadataAccountKey": "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==",
     "MetadataContainer": "metadata",
-    "MetadataBlobEndpoint": "http://127.0.0.1:10000/devstoreaccount1",
+    "MetadataBlobEndpoint": "https://pocaccessmanagement.blob.core.windows.net/",
     "DelegationsAccountName": "devstoreaccount1",
     "DelegationsAccountKey": "Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==",
     "DelegationsContainer": "delegationpolicies",
-    "DelegationsBlobEndpoint": "http://127.0.0.1:10000/devstoreaccount1",
+    "DelegationsBlobEndpoint": "https://pocaccessmanagement.blob.core.windows.net/",
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/appsettings.json:30]: Dev storage account endpoints should be removed

File: backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs [hunk 22]
```diff
@@ -430,7 +442,11 @@ public async Task Post_DeletePolicies_Sucess()
             HttpResponseMessage response = await _client.PostAsync("authorization/api/v1/delegations/DeletePolicy", content);
 
             string responseContent = await response.Content.ReadAsStringAsync();
-            List<Rule> actual = (List<Rule>)JsonConvert.DeserializeObject(responseContent, typeof(List<Rule>));
+            var options = new JsonSerializerOptions
+            {
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.AuthorizationAdmin.Tests/DelegationsControllerTest.cs:446]: If we actually need so specify this option for each test, then it move it to a common private property in the class, instead of having to set it up in each test

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 23]
```diff
@@ -230,17 +248,25 @@ private async Task<bool> WriteDelegationPolicyInternal(string policyPath, List<R
                         return false;
                     }
 
+                    ServiceResource resource = null;
+                    if (!string.IsNullOrWhiteSpace(resourceRegistryId))
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:252]: This has already been retrieved in the
if (resourceMatchType == ResourceAttributeMatchType.ResourceRegistry) scope (line 160)
and can be removed

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 24]
```diff
@@ -195,7 +204,16 @@ private async Task<bool> WriteDelegationPolicyInternal(string policyPath, List<R
                 try
                 {
                     // Check for a current delegation change from postgresql
-                    DelegationChange currentChange = await _delegationRepository.GetCurrentDelegationChange($"{org}/{app}", offeredByPartyId, coveredByPartyId, coveredByUserId);
+                    DelegationChange currentChange = new DelegationChange();
+
+                    string altinnAppId = string.Empty;
+                    if (!string.IsNullOrWhiteSpace(org) || !string.IsNullOrWhiteSpace(app))
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:210]: This can be moved to the if(resourceMatchType == ResourceAttributeMatchType.AltinnApp) scope earlier

Comment (Sondre-B) [reply backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:210]: So you mean moving also the code that uses the altinnAppId out of this scope? For the only code that actually need to be in the if (leaseId != null) scope is the one that depends on the bloblease

File: backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs [hunk 25]
```diff
@@ -230,17 +248,25 @@ private async Task<bool> WriteDelegationPolicyInternal(string policyPath, List<R
...
+                        AltinnAppId = altinnAppId,
                         OfferedByPartyId = offeredByPartyId,
                         CoveredByPartyId = coveredByPartyId,
                         CoveredByUserId = coveredByUserId,
                         PerformedByUserId = delegatedByUserId,
                         BlobStoragePolicyPath = policyPath,
-                        BlobStorageVersionId = blobResponse.Value.VersionId
+                        BlobStorageVersionId = blobResponse.Value.VersionId,
+                        ResourceId = resourceRegistryId != null ? resourceRegistryId : string.Empty,
+                        ResourceType = resource != null ? resource.ResourceType.ToString() : string.Empty
```
Comment (jonkjetiloye) [inline backend/src/Altinn.Authorizationadmin/Altinn.Authorizationadmin/Services/Implementation/PolicyAdministrationPoint.cs:269]: For app delegations when resourceRegistryId and ResourceType is null then they should be set to null in the database also. not empty string.

**Review by jonkjetiloye — APPROVED** [review]
